git-branchless-lib/src/git/object.rs

```diff
@@ -291,26 +293,59 @@ impl<'repo> Commit<'repo> {
 
     /// Amend this existing commit.
     /// Returns the OID of the resulting new commit.
-    #[instrument]
+    #[instrument(skip(signer))]
     pub fn amend_commit(
         &self,
-        update_ref: Option<&str>,
+        repo: &'repo Repo,
         author: Option<&Signature>,
         committer: Option<&Signature>,
         message: Option<&str>,
         tree: Option<&Tree>,
+        signer: Option<&dyn Sign>,
     ) -> Result<NonZeroOid> {
-        let oid = self
-            .inner
-            .amend(
-                update_ref,
-                author.map(|author| &author.inner),
-                committer.map(|committer| &committer.inner),
-                None,
-                message,
-                tree.map(|tree| &tree.inner),
-            )
-            .map_err(Error::Amend)?;
+        let parents = self.get_parents();
+        let parents = parents.iter().map(|parent| &parent.inner).collect_vec();
+
+        let new_author = match author {
+            Some(author) => &author.inner,
+            None => &self.inner.author(),
+        };
+        let new_committer = match committer {
+            Some(committer) => &committer.inner,
+            None => &self.inner.committer(),
+        };
+        let new_message = match message {
+            Some(message) => message,
+            None => match std::str::from_utf8(self.inner.message_bytes()) {
+                Ok(message) => message,
+                Err(e) => {
+                    error!(
+                        ?message,
+                        ?e,
+                        "failed to decode git commit message, not valid UTF-8"
+                    );
+                    return Err(Error::DecodeUtf8 { item: "message" });
+                }
+            },
+        };
+        let new_tree = match tree {
+            Some(tree) => &tree.inner,
+            None => &self.inner.tree().map_err(|err| Error::FindTree {
+                source: err,
+                oid: self.inner.tree_id().into(),
+            })?,
+        };
```

Comment:
[**question**] Why these changes? Or were these tagging along from the rebased implementation? At a glance, it seems like we may be able to use the previous versions (eg `author.map(|author| &author.inner)`, etc) but I haven't tested that.

This seems to introduce a break in behaviour: previously, passing `None` into this fn for any of these values would have passed that directly to `git2::amend()`, but here we seem to be discarding `None` and instead converting it to the existing author/message/tree/etc. Is that needed for signing? If not, can you see if this could be simplified and/or if any of these defaults are needed?

Reply:
Please disregard this comment. After sleeping on it, I realized why we need these here.